File: src/ollama.cpp
#include "ai_git/ollama.h"
#include <curl/curl.h>
#include <cstring>
#include <sstream>
#include <stdexcept>
#include <string>

namespace ai_git {
namespace ollama {

namespace {

size_t write_callback(void* contents, size_t size, size_t nmemb, void* userp) {
    ((std::string*)userp)->append((char*)contents, size * nmemb);
    return size * nmemb;
}

std::string escape_json(const std::string& s) {
    std::ostringstream o;
    for (char c : s) {
        switch (c) {
            case '"': o << "\\\""; break;
            case '\\': o << "\\\\"; break;
            case '\n': o << "\\n"; break;
            case '\r': o << "\\r"; break;
            case '\t': o << "\\t"; break;
            default: o << c;
        }
    }
    return o.str();
}

} // namespace

std::string generate_commit_message(const std::string& diff) {
    if (diff.empty()) {
        throw std::runtime_error("No staged changes to generate commit message for");
    }

    CURL* curl = curl_easy_init();
    if (!curl) {
        throw std::runtime_error("Failed to initialize CURL");
    }

    std::string response;
    std::string prompt = R"(Generate a concise, descriptive commit message for these changes. "
"Output only the commit message, no prefix like 'feat:' or 'fix:', no explanations. "
"Keep it under 72 characters if possible. "
"Changes:
)" + diff;

    std::string json_body = R"({"model": "llama3.2:3b", "prompt": ")" + escape_json(prompt) + R"(", "stream": false})";

    curl_easy_setopt(curl, CURLOPT_URL, "http://localhost:11434/api/generate");
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_body.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response);
    curl_easy_setopt(curl, CURLOPT_TIMEOUT, 120L);

    struct curl_slist* headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

    CURLcode res = curl_easy_perform(curl);
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);

    if (res != CURLE_OK) {
        throw std::runtime_error("Failed to connect to Ollama. Is it running?");
    }

    std::string key = "\"response\":\"";
    size_t start = response.find(key);
    if (start == std::string::npos) {
        throw std::runtime_error("Invalid response from Ollama");
    }
    start += key.size();
    size_t end = response.find("\"}", start);
    if (end == std::string::npos) {
        end = response.find("\"", start);
    }
    std::string msg = response.substr(start, end - start);

    size_t pos;
    while ((pos = msg.find("\\n")) != std::string::npos) {
        msg.replace(pos, 2, "\n");
    }
    while ((pos = msg.find("\\\"")) != std::string::npos) {
        msg.replace(pos, 2, "\"");
    }

    if (!msg.empty() && msg.back() == '"') {
        msg.pop_back();
    }

    return msg;
}

} // namespace ollama
} // namespace ai_git

File: lib/ai_git/ollama.rb
# frozen_string_literal: true

require 'net/http'
require 'uri'
require 'json'

module AIGit
  module Ollama
    module_function

    def escape_json(string)
      string.gsub('\\', '\\\\')
            .gsub('"', '\"')
            .gsub("\n", '\\n')
            .gsub("\r", '\\r')
            .gsub("\t", '\\t')
    end

    def generate_commit_message(diff, model_name)
      raise 'No staged changes to generate commit message for' if diff.to_s.strip.empty?

      prompt = <<~PROMPT
        You are an expert Git commit message writer. Output ONLY the commit message — no explanations, no markdown, no backticks, no preamble.

        Here are the changes:
        #{diff}

        STRICT OUTPUT FORMAT (follow exactly):

        <short imperative title, max 72 chars>

        <blank line>

        ## Summary
        <2–4 bullet points covering the most important changes. Each bullet starts with a verb.>

        ## Why
        <1–3 sentences explaining the motivation or context behind the change. Omit if the reason is obvious.>

        RULES:
        - Title line: short, specific, imperative mood (e.g. "Add JWT login with refresh token support"). Avoid vague titles like "Update stuff" or "Fix bug".
        - Summary bullets: describe WHAT changed, not HOW the code looks. Focus on behaviour and impact.
        - Why section: explain the problem being solved or the goal being achieved. Skip if it adds no value.
        - No filler phrases ("this commit", "this PR", "as per discussion").
        - No line should exceed 72 characters.

        EXAMPLES OF GOOD OUTPUT:

        Add JWT-based login with refresh token support

        ## Summary
        - Implement login endpoint with access and refresh token issuance
        - Add token refresh route with rotation and expiry validation
        - Protect private routes via middleware that verifies access tokens
        - Store refresh tokens using encrypted HTTP-only cookies

        ## Why
        Users were being logged out on every page reload. Refresh tokens allow
        sessions to persist securely without requiring re-authentication.

        ---

        Prevent nil crash when user preferences are missing

        ## Summary
        - Add nil guard in ReportGenerator#process before accessing preferences
        - Fall back to system defaults when preferences object is absent

        ## Why
        Reports were raising NoMethodError in production for users created
        before the preferences feature shipped.

        ---

        Now generate the commit message:
      PROMPT

      json_body = {
        model: model_name,
        prompt: prompt,
        stream: false,
        # These parameters help a lot with strictness:
        temperature: 0.3,
        top_p: 0.9,
        stop: ["\n\n\n", '```', 'Here is', 'The commit message'],
        num_predict: 400 # Limit output length
      }.to_json

      uri = URI('http://localhost:11434/api/generate')
      request = Net::HTTP::Post.new(uri)
      request['Content-Type'] = 'application/json'
      request.body = json_body

      response = Net::HTTP.start(uri.host, uri.port, read_timeout: 120) do |http|
        http.request(request)
      end

      raise 'Failed to connect to Ollama. Is it running?' unless response.is_a?(Net::HTTPSuccess)

      data = JSON.parse(response.body)
      message = data['response'].to_s.strip

      # Aggressive cleaning
      message = message.gsub(/^(Here is|The commit message is|```|json|markdown)/i, '')
                       .gsub(/^>\s*/, '')
                       .gsub(/\\n/, "\n")
                       .strip

      # Remove common unwanted prefixes/suffixes
      lines = message.lines.map(&:strip)
      lines.reject! { |line| line.match?(/^(Here|Output|Generated|Based on|The changes)/i) }

      message = lines.join("\n").strip

      # Ensure it has at least a title
      if message.lines.count < 1 || message.strip.empty?
        message = 'chore: update code' # fallback
      end

      message
    end
  end
end
